<a href="https://colab.research.google.com/github/gandhias-a11y/colab/blob/main/Lab05_jpynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Q1

In [ ]:
import numpy as np
a = np.array([1, 2, 3, 4, 5])
print(a)

[1 2 3 4 5]


Q2

In [ ]:
import numpy as np
def what(n):
  a = np.array([1, 2, 3, 4], ndmin=n)
  print(a.ndim)
  return a
print(what(1))
print(what(2))
print(what(3))

1
[1 2 3 4]
2
[[1 2 3 4]]
3
[[[1 2 3 4]]]


The what() function outputs a particular member of the array followed by the array itself encased in a number of brackets that is the same as the ndmin input to reflect the change in dimension that the function created.

Q3

In [ ]:
import numpy as np
a = np.array([1, 2, 3, 4], ndmin = 2)
print(a)

[[1 2 3 4]]


Q4

In [ ]:
import numpy as np

a = np.array([1, 2, 3, 4])
print(a.ndim)

b = np.array([[1, 2], [3, 4]])
print(b.ndim)

1
2


Q5

In [ ]:
import numpy as np
a = np.array([1, 2, 3, 4])
print(a[0])

1


Q6

In [ ]:
import numpy as np
a = np.array([10, 20, 30, 40, 50])
print(a[4])

50


Q7

In [ ]:
import numpy as np
a = np.array([[10, 20, 30, 40], [50, 60, 70, 80]])
print(a[1][0] )

50


Q8

In [ ]:
import numpy as np
a = np.array([10, 20, 30, 40, 50])
print( a[-1])

50


Q9

In [ ]:
import numpy as np
a = np.array([10, 15, 20, 25, 30, 35, 40])
print(a[1:4])

[15 20 25]


Q10

In [ ]:
import numpy as np
a = np.array([10, 15, 20, 25, 30, 35, 40])
print(a[2:4])

[20 25]


Q11

In [ ]:
import numpy as np
a = np.array([10, 15, 20, 25, 30, 35, 40])
print(a[1:4:2])

[15 25]


Q12

In [ ]:
import numpy as np
a = np.array([10, 15, 20, 25, 30, 35, 40])
print(a[::2])

[10 20 30 40]


Q13

In [ ]:
import numpy as np
a = np.array([1, 2, 3])
print(a.dtype)

int64


Q14

In [ ]:
import numpy as np
a = np.array([1, 2, 3, 4], dtype= 'S')

print(a.dtype)
print(a[0] + a[1])
a = np.array([5, 1, -1, 3], dtype='S')
b = a.astype('i')
print(b[0] + b[1] + b[3])
print(a[0] + a[1] + a[3])

|S1
b'12'
9
b'513'


My prediction for the outputs above was that they would reflect the data type in an unfamiliar syntax and then would attempt to either add or concatenate a and b depending on their current types. In the above expressions, the result |S1 means that the output is a string, the result b'12' means that the array is being converted back to an integer, the result of 9 is the sum of our now integer values, and the final result of b'513' is a concatenation of string values rather than integer addition.

Q15

In [ ]:
a = np.array([1.1, 2.1, 3.1])
b = a.astype("i")
print(b)

[1 2 3]


Q16

In [ ]:
a = np.array([1, 2, 3, 4, 5])
x = a.copy()
print(x)

x[0] = -5
print(a)
b = a
b[0] = -25
print(a)

[1 2 3 4 5]
[1 2 3 4 5]
[-25   2   3   4   5]


The code does not change with the first expression because we have changed a copy of a, not something directly connected to a itself. The second print statement does change a because the initialization of b involved directly connecting it to a.

Q17

In [ ]:
a = np.array([1, 2, 3, 4, 5])
d = a.view()

print(a.base)
print(d.base)
b = a
print(b.base)
c = a.copy()
print(c.base)

None
[1 2 3 4 5]
None
None


The only expression that prints something is the one that involves the view version. The base function does not work on regular array types.

Q18

In [ ]:
a = np.array([1, 2, 3, 4])
print(a.shape)
a.reshape(1, 4)
print(a)
a.reshape(2, 2)
print(a)
a.reshape(4, 1)
print(a)

(4,)
[1 2 3 4]
[1 2 3 4]
[1 2 3 4]


The output of the first expression shape() just tells us that the array is a four by one. The reshapes do not ultimately change the array's dimensions which is why the the result of the print functions are the same (4x1, 1x4, and 2x2 all equal 4).

Q19

In [ ]:
a = np.array([1, 2, 3, 4])
a= a.reshape(2,2)
print(a)

a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12])
b = a.reshape(4, 3)
print(b)
b = a.reshape(3, 4)
print(b)

[[1 2]
 [3 4]]
[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


The reshaping of the new a array reconfigures the structure of the array to have x rows and y columns with the use of the function reshape(x,y). The first reshape in b shifts the array to have four rows and three columns. The second reshape shifts it to have 3 rows and 4 columns.

Q20

In [ ]:
a = np.array([[1, 2, 3, 4, 5, 6], [7, 8, 9, 10, 11, 12]])
b = a.reshape(12,)
print(b)

[ 1  2  3  4  5  6  7  8  9 10 11 12]


Q21

In [6]:
import numpy as np
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
c =np.concatenate((a, b))
print(c)

a = np.array([[1, 2], [3, 4]])
b = np.array([[5, 6], [7, 8]])
c = np.concatenate((a, b), axis=1)
print (c)
c  = np.stack((a, b), axis = 0)
print(c)

[1 2 3 4 5 6]
[[1 2 5 6]
 [3 4 7 8]]
[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


The first print statement of c results in consolidation of arrays a and b into one array that is split into 2 rows because the concatenate function specified the axis equaled 1.
The second print statement separates c into two 2x2 2-d arrays stacked to form a 3-d array.

Q22

In [ ]:
a = np.array([1, 2, 3, 4, 5, 4, 4])
where = np.where(a==4)
print(where)

(array([3, 5, 6]),)


Q23

In [ ]:
a = np.array([3, 2, 0, 1])
b = np.sort(a)
print(b)

[0 1 2 3]


Q24

In [ ]:
## Dependencies
import numpy as np
## Data: yearly salary in ($1000) [2017, 2018, 2019]
alice = [99, 101, 103]
bob = [110, 108, 105]
tim = [90, 88, 85]
names = ['alice', 'bob', 'tim']
salaries = np.array([alice, bob, tim])
taxation = np.array([[0.2, 0.25, 0.22],
 [0.4, 0.5, 0.5],
 [0.1, 0.2, 0.1]])
## One-liner
max_income = np.max(salaries - salaries * taxation)

## Result
print(max_income)

#  pull the name of the person with the highest salary from the matrix by using this idea of selective Boolean indexing. Problem recap: How do we find the person with maximum after-tax income in a group of  people, given their yearly salary and tax rates?
after_tax_incomes = (salaries - salaries * taxation).max(axis=1)
print(after_tax_incomes)
max_index = np.argmax(after_tax_incomes)
print(max_index)
print(names[max_index])

81.0
[80.34 66.   81.  ]
2
tim


Q25

In [8]:
a = np.array([[[1, 2], [3, 4]],
              [[5, 6], [7, 8]]])
print(np.average(a, axis=0))
print(np.average(a, axis=1))
print(np.average(a, axis=2))

[[3. 4.]
 [5. 6.]]
[[2. 3.]
 [6. 7.]]
[[1.5 3.5]
 [5.5 7.5]]


The three different print statements result in the average of the array's contents across its different axes. The first statement averages the values vertically in their current shape (ex: the average of 1 and 5 is 3 which is why 3 is the first value printed).
The second print statement averages the values across the first entry of each of the 1x1 arrays (ex: the average of 1 and 3 is 2 which is why 2 is the first value printed).
The third print statement averages the values within each 1x1 array in a (ex: the average of 1 and 2 is 1.5 which is why 1.5 is what is printed first).

Q26

In [11]:
a = np.array([[[ 1,  2,  3], [22,  5,  6], [ 7,  8,  9]],
              [[11, 12, 10], [ 4, 14, 15], [16, 17, 18]],
              [[19, 20, 21], [13, 23, 24], [25, 26, 27]]])

b = np.sort(a, axis=0)
print(b)
print()
c = np.sort(a, axis=1)
print(c)
print()
d = np.sort(a, axis=2)
print(d)

[[[ 1  2  3]
  [ 4  5  6]
  [ 7  8  9]]

 [[11 12 10]
  [13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]
  [25 26 27]]]

[[[ 1  2  3]
  [ 7  5  6]
  [22  8  9]]

 [[ 4 12 10]
  [11 14 15]
  [16 17 18]]

 [[13 20 21]
  [19 23 24]
  [25 26 27]]]

[[[ 1  2  3]
  [ 5  6 22]
  [ 7  8  9]]

 [[10 11 12]
  [ 4 14 15]
  [16 17 18]]

 [[19 20 21]
  [13 23 24]
  [25 26 27]]]


The first sort function arranges the values of the first column of array in ascending numerical order. This is why 4, 22, and 13 all switch places.
The second sort function rearranged the 1x3 arrays to make sure that the first values of each array are in order. That is why 7 and 22 switched places, for example, because 1, 7, and 22 had to be placed in ascending order.
The third sort function arranges the values within the 1x3 arrays to be in ascending order. For example, [22, 5, 6] switches to [5, 6, 22].
